audio-echo/app/src/main/java/com/google/sample/echo/MainActivity.java

```diff
@@ -120,11 +133,46 @@ private void queryNativeAudioParameters() {
         nativeSampleFormat ="";    //TODO: find a way to get the native audio format
     }
     private void updateNativeAudioUI() {
-        status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
-                "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
-                "nativeSampleFormat  = " + nativeSampleFormat);
+        if(permissionGranted) {
+            status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
+                    "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
+                    "nativeSampleFormat  = " + nativeSampleFormat);
+        } else {
+            status_view.setText("Error: Resource Permission not Granted" +
```

Comment:
+1, Ideally this should be done just before the permission request. If you are restructuring this change based on my suggestion (see above), a call to shouldShowRequestRationale might not be necessary though.